### ISSUE TITLE
How to efficiently implement block operators?

### ISSUE THREAD

**Issue by mehrhardt** [odlgroup/odl#1349]
What is the best way to implement block operators, i.e. operators that consist of a matrix of operators. In the ODL example one finds two different notations with a very different numerical performance. Here I implemented both for symmetrized gradients related to TGV type methods.

It turns out that the evaluation of `E` is twice as fast on my machine compared to `E2`. This comes in addition to the in my view "nicer" code. Is this now the recommended way of writing these in ODL? If so, one probably should remove the other notation from the "simple" examples?

```
import odl

X = odl.uniform_discr([0, 0, 0], [1, 1, 1], [200, 200, 60])

PD = odl.BroadcastOperator(
        *[odl.PartialDerivative(X, i, method='backward', pad_mode='symmetric')
          for i in range(3)])

E = odl.operator.ProductSpaceOperator(
    [[PD[0], 0, 0],
     [0, PD[1], 0],
     [0, 0, PD[2]],
     [PD[1], PD[0], 0],
     [PD[2], 0, PD[0]],
     [0, PD[2], PD[1]]])

P = odl.operator.BroadcastOperator(
        *[odl.ComponentProjection(PD.range, i) for i in range(3)])

E2 = odl.operator.BroadcastOperator(
        PD[0] * P[0],
        PD[1] * P[1],
        PD[2] * P[2],
        PD[1] * P[0] + PD[0] * P[1],
        PD[2] * P[0] + PD[0] * P[2],
        PD[2] * P[1] + PD[1] * P[2])

v = odl.phantom.white_noise(PD.range)

Ev = E.range.element()
E2v = E2.range.element()

%timeit Ev = E(v)
%timeit E2v = E2(v)
%timeit E(v, out=Ev)
%timeit E2(v, out=E2v)
```

On the note of writing nice code. It would be nice to write the above code in a more concise notation like

`G = odl.Gradient(X)`
then 
`G[0] = odl.PartialDerivative(X, 0)`

and 
```
V = X ** 3
I = odl.IdentityOperator(V)
```
then
`I[0] = odl.ComponentProjection`(V, 0)

**Comment by kohr-h** [comment]
The difference in runtime is to be expected, since `E` has 9 operator evaluations, while `E2` has 18. With the sizes you chose, a partial derivative is likely not much more work than a copy (which is what `ComponentProjection` effectively does).

> then
> G[0] = odl.PartialDerivative(X, 0)

I assume this would mean picking an operator block rather than setting it; the latter is certainly not something we'd do.

There are two main problems with the picking of components from an `IdentityOperator` or a `Gradient`:

1. They can mean different things. In one case, you pick a domain component (reductions), in the other case a range component (broadcasting). And there's a third case where both domain and range are product spaces. What would the above `I[0]` mean, for instance? It's ambiguous and would require a convention, e.g., rows first.
2. It's not obvious how to rip apart operators in general. This happens to work for `Gradient` and `Identity`, or other simple things like `MultiplyOperator`. But for other operators, it's way less obvious, and I don't want to introduce a specific API just for a set of operators that support it, given that another, slightly less succinct construct already gives you this functionality.
   So I tend to think that if indexing of an operator is desired, one should explicitly create a row vector (`ReductionOperator`), column vector (`BroadcastOperator`) or matrix of operators (`ProductSpaceOperator`) and use those for all kinds of indexing purposes.
   
So in your example, I think the `E` variant is the most compact way of expressing your intention (apart from the fact that you don't need the `BroadcastOperator`, you can just work with the plain list).

> Is this now the recommended way of writing these in ODL? If so, one probably should remove the other notation from the "simple" examples?

I don't quite know. Both are rather advanced constructs, so I tend to think that we should go with the less verbose one, which would be the `ProductSpaceOperator` in cases like this one.

**Comment by mehrhardt** [comment]
Thanks @kohr-h for the answer. Why does `ComponentProjection` need to make a copy? That sounds unnecessarily inefficient to me but maybe I am overlooking some details. I actually expected that `ComponentProjection` would not compute / copy anything, thus my surprise in runtime. Does the `IdentityOperator` copy something?

Regarding the slicing of operators, I indeed meant the mathematical "=" in `G[0] = odl.PartialDerivative(X, 0)`. Such a slicing notation does not make much sense in general if the operator `G` has no special structure. Ripping the operator apart by column is possible but kind of strange. In this case, one would have `G = \sum_i G[i]` and indeed it is not clear for which operators this makes sense apart from `SeparableSum` or the `ReductionOperator` as this is kind of the definition of it). However, ripping an operator apart row wise does make sense for every operator that maps into a product space, such as the gradient and the identity if the domain is a product space. Thus, one could introduce this notation for every `ProductSpaceOperator` without any significant cause of confusion, no?

**Comment by kohr-h** [comment]
> Thanks @kohr-h for the answer. Why does ComponentProjection need to make a copy? That sounds unnecessarily inefficient to me but maybe I am overlooking some details. I actually expected that ComponentProjection would not compute / copy anything, thus my surprise in runtime. Does the IdentityOperator copy something?

Yes, also `IdentityOperator` usually makes a copy. The reason for this is that we don't allow the input to be mutated, except in the case when the user explicitly passes the input as the `out` parameter as well. So if you do
```python
I = odl.IdentityOperator(odl.rn(2))
x = I.domain.one()
I(x, out=x)
```
then this is actually a no-op. Of course, this only works if domain and range are the same, and it won't cover intermediate operators in a composition.

This is a limitation of ODL which could definitely be interesting looking into. For example, I could imagine something like `OperatorComp(A, B, allow_inplace=True)` for optimization, where temporary arrays are modified in place whenever possible. The question is how far you would drive this, for instance, should one try to optimize a composition mapping like `X -> Y -> X -> Y` in such a way that the temporaries are re-used across the whole chain? That could also be quite counterproductive in terms of memory consumption.

For the slicing, I agree that indexing along the row coordinate always makes sense for operators whose range is a product space, without confusion. It's in line with how indexing everywhere else works, including NumPy arrays. You can easily have a default implementation that simply runs the whole thing and then picks the component in question. For certain operators there is an obvious way how to construct the sliced operators, e.g. gradient or identity.

My question is rather: should we do it? My concern is mostly from a maintenance and user expectation point of view. A construct like this creates more cross-links between classes in ODL, which brings with it the burden of having to always keep the operators in sync. Users would expect (and rightly so) that `G[0](x) == G(x)[0]`, and we would have to guarantee that. For example, we couldn't add a feature to `Gradient` but not to `PartialDerivative` if we think that `Gradient` is the much more important class.
Furthermore, unless we put in quite a bit of work, we'd have a really crappy default implementation for most operators that would likely annoy users.
This is what we need to weigh this against the benefit, which is a nicer way of doing something that's already possible.

**Comment by mehrhardt** [comment]
Sure, I fully understand. I am also divided on this; perhaps it is better not to have this functionality.

> This is a limitation of ODL which could definitely be interesting looking into. For example, I could imagine something like OperatorComp(A, B, allow_inplace=True) for optimization, where temporary arrays are modified in place whenever possible. The question is how far you would drive this, for instance, should one try to optimize a composition mapping like X -> Y -> X -> Y in such a way that the temporaries are re-used across the whole chain? That could also be quite counterproductive in terms of memory consumption.

I am not sure I fully understand where you are getting at. Why would reusing temporaries increase the memory consumption? 

It is interesting that you mention that operator composition does not allow inplace computations. I was actually assuming that ODL would do that. Why would ODL not do that?

**Comment by kohr-h** [comment]
> I am not sure I fully understand where you are getting at. Why would reusing temporaries increase the memory consumption?

Because in the worst case, you would have to keep them alive for the whole chain of operators. Imagine an operator chain `X -> Y -> Z -> lots of stuff -> Z -> Y -> X`. With reusable temporaries, you'd need the temporaries for `X`, `Y` and `Z` across a long span of time while doing other stuff. The current "greedy" approach is the complete opposite.


> It is interesting that you mention that operator composition does not allow inplace computations. I was actually assuming that ODL would do that. Why would ODL not do that?

You can do that with the input and output at both ends of the chain by using `op_comp(x, out=x)` if it makes sense, but in the middle it's not so obvious how to achieve the same using a given operator composition.
There is a possibility, however, to do this by the (little known) extra parameters to `OperatorComp`. If you have a chain of operators all mapping `X -> X` you could do a sneaky thing like this:
```python
A, B, C = ...  # some operators with domain == range
tmp = A.range.element()
BA = OperatorComp(B, A, tmp=tmp)
CBA = OperatorComp(C, BA, tmp=tmp)
```
In this case, no extra copies would be made. (Note that I was tempted to use `x` as temporary, but that will give incorrect results). This is clearly not as concise and elegant as writing `CBA = C * B * A`.
